### PR TITLE
update RTree Spatial Index triggers

### LINKED ISSUE #639: RTree _update1 trigger not compatible of UPSERT capability
I've found an issue when wanting to use the UPSERT capability on a table with a RTree.
For example the following UPSERT that inserts a FID 1 or update it
```
INSERT INTO foo (fid, geom) VALUES (1, put_here_some_gpkg_geometry_blob) on conflict (fid) DO UPDATE SET geom=excluded.geom;
```
fails with an error like "UNIQUE constraint failed: foo_rtree.id (19)" when updating an existing row.
I've spotted it to be caused by the "_update1" trigger that has a "INSERT OR REPLACE INTO" statement.
More details in https://sqlite.org/forum/forumpost/8c8de6ff91 where the SQLite author answered that SQLite3 behave as documented here, although a bit surprising.
In https://github.com/OSGeo/gdal/pull/6532, I've put a workardound where I remove the _update1 trigger and replace it with 2 separate triggers to avoid the need of a "INSERT OR REPLACE INTO" statement.

One is a _update1_old_geom_notnull trigger that does an update of the rtree when both the old and new geoms are not null/empty
```
pszSQL = sqlite3_mprintf(
                   "CREATE TRIGGER \"%w_update1_old_geom_notnull\" AFTER UPDATE OF \"%w\" ON \"%w\" "
                   "WHEN OLD.\"%w\" = NEW.\"%w\" AND "
                   "(NEW.\"%w\" NOTNULL AND NOT ST_IsEmpty(NEW.\"%w\") AND OLD.\"%w\" NOTNULL AND NOT ST_IsEmpty(OLD.\"%w\")) "
                   "BEGIN "
                   "UPDATE \"%w\" SET "
                   "minx = ST_MinX(NEW.\"%w\"), maxx = ST_MaxX(NEW.\"%w\"),"
                   "miny = ST_MinY(NEW.\"%w\"), maxy = ST_MaxY(NEW.\"%w\") "
                   "WHERE id = NEW.\"%w\";"
                   "END",
                   osRTreeName.c_str(), pszC, pszT,
                   pszI, pszI,
                   pszC, pszC, pszC, pszC,
                   osRTreeName.c_str(),
                   pszC, pszC,
                   pszC, pszC,
                   pszI);
```

and one is a _update1_old_geom_null that does an insert in the rtree when the old geom is null/empty and the new one is not:
```
    pszSQL = sqlite3_mprintf(
                   "CREATE TRIGGER \"%w_update1_old_geom_null\" AFTER UPDATE OF \"%w\" ON \"%w\" "
                   "WHEN OLD.\"%w\" = NEW.\"%w\" AND "
                   "(NEW.\"%w\" NOTNULL AND NOT ST_IsEmpty(NEW.\"%w\") AND (OLD.\"%w\" ISNULL OR ST_IsEmpty(OLD.\"%w\"))) "
                   "BEGIN "
                   "INSERT INTO \"%w\" VALUES ("
                   "NEW.\"%w\","
                   "ST_MinX(NEW.\"%w\"), ST_MaxX(NEW.\"%w\"),"
                   "ST_MinY(NEW.\"%w\"), ST_MaxY(NEW.\"%w\")"
                   "); "
                   "END",
                   osRTreeName.c_str(), pszC, pszT,
                   pszI, pszI,
                   pszC, pszC, pszC, pszC,
                   osRTreeName.c_str(),
                   pszI,
                   pszC, pszC,
                   pszC, pszC);
```

Should the spec be updated to replace the _update1 trigger by the 2 above ones ?

### LINKED ISSUE #660: Be more robust about replacing spatial index triggers.
#414 revealed a broken `update3` trigger that was fixed in #415. #639 revealed a broken `update1` trigger that is scheduled to be fixed as part of #657. However, the infrastructure currently lacks a clear ability to advertise whether fixing the triggers has occurred. I originally proposed a separate extension to reveal the fix for #639. However, it is now clear to me that we need a way to advertise the presence of both fixes in a way that is usable all the way back to GeoPackage 1.0.

My new proposal is to do the following:
* Deprecate (drop) the `update1` and `update3` triggers entirely
* Add the `update5` trigger (replacement for `update3`) and `update6` and `update7` triggers (replacement for `update1`)

This way, any GeoPackage client can look for the presence of the flawed triggers 1 and 3 and absence of the corrected triggers 5, 6, and 7 and know to fix the issue if needed. This can be done without introducing a new extension. We can just make it part of the revised Annex F.3 in GeoPackage 1.4.

### PR DESCRIPTION
**Pull request by jyutzler**
Closes #660
Closes #639

### CODE REVIEW

**Comment by cnreediii** [comment]
@jyutzler Looks fine to me. Two minor suggestions: 1.) Provide a link to a R-Tree info page (Wikipedia?) and 2.) if referencing R-Trees as a technology and not a product/extension/tablename, the spelling as per Guttman is R-tree. Cheers

**Comment by jyutzler** [comment]
Carl,
I went ahead and made that change. I'm not going to cite Wikipedia in a standard document so I added Guttman to the bibliography.
Anyone else? Anyone inclined to make a motion?
-Jeff
P.S. No meeting today, let's just follow up on this off-line.
________________________________

**Comment by cnreediii** [comment]
@jyutzler Thanks. No issue with just using Guttman as a reference.

File: spec/core/release_notes/1.4.0/clause-6-substantive.adoc [hunk 1]
```diff
@@ -19,3 +19,22 @@ If anything, this requirement was just causing confusion.
 The presence of extensions would cause the Executable Test Suite (ETS) to skip Requirement 4, which has implications in how ETS results are interpreted.
 
 This change has no significant operational impact.
+
+=== Update R-tree Index Triggers
+Two changes have been made to the R-tree Spatiial Index Triggers,
```
Comment (heidivanparys) [inline spec/core/release_notes/1.4.0/clause-6-substantive.adoc:24]: Typo
```suggestion
Two changes have been made to the R-tree Spatial Index Triggers,
```

File: spec/core/annexes/extension_spatialindex.adoc [hunk 2]
```diff
@@ -191,8 +229,8 @@ Definition of SQLite configuration settings
 [cols=",,,",options="header"]
 |======
 |Setting compile or runtime |Option |Shall / Not (Value) | Discussion
-|compile |SQLITE_ENABLE_RTREE |Shall |RTrees ares used for GeoPackage Spatial Indexes
-|compile |SQLITE_RTREE_INT_ONLY |Not |RTrees with floating point values are used for GeoPackage spatial indexes
+|compile |SQLITE_ENABLE_RTREE |Shall |R-trees ares used for GeoPackage Spatial Indexes
```
Comment (heidivanparys) [inline spec/core/annexes/extension_spatialindex.adoc:232]: Typo
```suggestion
|compile |SQLITE_ENABLE_RTREE |Shall |R-trees are used for GeoPackage Spatial Indexes
```

File: spec/core/annexes/extension_spatialindex.adoc [hunk 3]
```diff
@@ -1,10 +1,10 @@
 [[extension_rtree]]
-=== RTree Spatial Indexes
+=== R-tree Spatial Indexes
 
 [float]
 ==== Introduction
 
-The RTree Spatial Indexes extension provides a means to encode an RTree index for geometry values in a GeoPackage. An RTree index provides a significant performance advantage for searches with basic envelope spatial criteria that return subsets of the rows in a feature table with a non-trivial number (thousands or more) of rows.^<<K26>>^
+The R-tree Spatial Indexes extension provides a means to encode an R-tree index for geometry values in a GeoPackage. An R-tree index<<B28>> provides a significant performance advantage for searches with basic envelope spatial criteria that return subsets of the rows in a feature table with a non-trivial number (thousands or more) of rows.^<<K26>>^
```
Comment (heidivanparys) [inline spec/core/annexes/extension_spatialindex.adoc:7]: Missing whitespace
```suggestion
The R-tree Spatial Indexes extension provides a means to encode an R-tree index for geometry values in a GeoPackage. An R-tree index <<B28>> provides a significant performance advantage for searches with basic envelope spatial criteria that return subsets of the rows in a feature table with a non-trivial number (thousands or more) of rows.^<<K26>>^
```

**Comment by jyutzler** [comment]
@heidivanparys I think 5d862cb1590057f74b61b5ae97758457d1c87267 covers everything you identified.

**Comment by rouault** [comment]
I believe there's an issue with the naming of _update6 and _update7 triggers. Fix proposal in https://github.com/opengeospatial/geopackage/pull/663

**Pull request merged**